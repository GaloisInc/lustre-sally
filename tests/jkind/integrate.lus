-- This file has 2 properties, both valid.

node integ(x : int) returns (sum : int);
let
  sum = x + (0 -> pre sum);
tel;

node main(x : int; y : int) returns (z : int);
var
  history : bool;
  prop1 : bool;
  prop2 : bool;
let
  z = integ(x);

  history = x > 0 and (true -> pre history);
  prop1 = history => z > 0;

  --%PROPERTY prop1;

  prop2 = integ(x) + integ(y) = integ(x + y);
  --%PROPERTY prop2;
tel;
