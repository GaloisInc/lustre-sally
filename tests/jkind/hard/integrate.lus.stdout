Core Lustre
===========

node (x_4 : int, y : int)
  returns (z)
  assumes ()
  shows (prop2, prop1)
let
  x_13 = x_4
  valof_pre_z = pre z
  valof_arr_18 = 0 -> valof_pre_z
  z = Add(x_13, valof_arr_18)
  valof_pre_history = pre history
  valof_arr_20 = true -> valof_pre_history
  history = And(Gt(x_4, 0), valof_arr_20)
  prop1 = Implies(history, Gt(z, 0))
  x_14 = x_4
  valof_pre___no_static_21 = pre __no_static_10
  valof_arr_22 = 0 -> valof_pre___no_static_21
  __no_static_10 = Add(x_14, valof_arr_22)
  x_15 = y
  valof_pre___no_static_23 = pre __no_static_11
  valof_arr_24 = 0 -> valof_pre___no_static_23
  __no_static_11 = Add(x_15, valof_arr_24)
  x_16 = Add(x_4, y)
  valof_pre___no_static_25 = pre __no_static_12
  valof_arr_26 = 0 -> valof_pre___no_static_25
  __no_static_12 = Add(x_16, valof_arr_26)
  prop2 = Eq(Add(__no_static_10, __no_static_11), __no_static_12)
tel
Sally Model
===========

(define-state-type
   S
   (
     (history Bool)
     (prop1 Bool)
     (prop2 Bool)
     (y Int)
     (z Int)
     (|__no_static:10| Int)
     (|__no_static:11| Int)
     (|__no_static:12| Int)
     (|gal-initializing| Bool)
     (|valof_arr:18:init| Bool)
     (|valof_arr:18| Int)
     (|valof_arr:20:init| Bool)
     (|valof_arr:20| Bool)
     (|valof_arr:22:init| Bool)
     (|valof_arr:22| Int)
     (|valof_arr:24:init| Bool)
     (|valof_arr:24| Int)
     (|valof_arr:26:init| Bool)
     (|valof_arr:26| Int)
     (|valof_pre___no_static:21| Int)
     (|valof_pre___no_static:23| Int)
     (|valof_pre___no_static:25| Int)
     (|valof_pre_history| Bool)
     (|valof_pre_z| Int)
     (|x:13| Int)
     (|x:14| Int)
     (|x:15| Int)
     (|x:16| Int)
     (|x:4| Int))
   (
     (y Int)
     (|x:4| Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |valof_arr:18:init| true)
      (= |valof_arr:20:init| true)
      (= |valof_arr:22:init| true)
      (= |valof_arr:24:init| true)
      (= |valof_arr:26:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|x:4| input.|x:4|)
      (= next.y input.y)
      (= next.|x:13| next.|x:4|)
      (= next.|valof_pre_z| state.z)
      (ite
         state.|valof_arr:18:init|
         (= next.|valof_arr:18| 0)
         (= next.|valof_arr:18| next.|valof_pre_z|))
      (= next.|valof_arr:18:init| false)
      (=
         next.z
         (+ next.|x:13| next.|valof_arr:18|))
      (= next.|valof_pre_history| state.history)
      (ite
         state.|valof_arr:20:init|
         (= next.|valof_arr:20| true)
         (= next.|valof_arr:20| next.|valof_pre_history|))
      (= next.|valof_arr:20:init| false)
      (=
         next.history
         (and
            (< 0 next.|x:4|)
            next.|valof_arr:20|))
      (=
         next.prop1
         (=>
            next.history
            (< 0 next.z)))
      (= next.|x:14| next.|x:4|)
      (= next.|valof_pre___no_static:21| state.|__no_static:10|)
      (ite
         state.|valof_arr:22:init|
         (= next.|valof_arr:22| 0)
         (= next.|valof_arr:22| next.|valof_pre___no_static:21|))
      (= next.|valof_arr:22:init| false)
      (=
         next.|__no_static:10|
         (+ next.|x:14| next.|valof_arr:22|))
      (= next.|x:15| next.y)
      (= next.|valof_pre___no_static:23| state.|__no_static:11|)
      (ite
         state.|valof_arr:24:init|
         (= next.|valof_arr:24| 0)
         (= next.|valof_arr:24| next.|valof_pre___no_static:23|))
      (= next.|valof_arr:24:init| false)
      (=
         next.|__no_static:11|
         (+ next.|x:15| next.|valof_arr:24|))
      (=
         next.|x:16|
         (+ next.|x:4| next.y))
      (= next.|valof_pre___no_static:25| state.|__no_static:12|)
      (ite
         state.|valof_arr:26:init|
         (= next.|valof_arr:26| 0)
         (= next.|valof_arr:26| next.|valof_pre___no_static:25|))
      (= next.|valof_arr:26:init| false)
      (=
         next.|__no_static:12|
         (+ next.|x:16| next.|valof_arr:26|))
      (=
         next.prop2
         (=
            (+ next.|__no_static:10| next.|__no_static:11|)
            next.|__no_static:12|))))

(query
   TS
   (or |gal-initializing| prop2))
(query
   TS
   (or |gal-initializing| prop1))

[Lustre] Validating properties:
[Lustre] Property prop2... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Property prop1... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Model status: [[1;92mValid[0m] 
